{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "def create_gcs_placeholder(output_folder):\n",
    "    # Create a placeholder to simulate folder structure in GCS\n",
    "    placeholder_file = f\"{output_folder}placeholder\"\n",
    "    subprocess.run(f\"gsutil cp /dev/null {placeholder_file}\", shell=True, check=True)\n",
    "\n",
    "\n",
    "def process_cell_line(cell_line, bam_location, transcript_index, output_bucket, local_output_folder, local_results_folder):\n",
    "    # Convert BAM to FASTQ\n",
    "    read1_fq = f\"{local_output_folder}reads_1.fq\"\n",
    "    read2_fq = f\"{local_output_folder}reads_2.fq\"\n",
    "\n",
    "    try:\n",
    "        print(f'Bam conversion for {cell_line} starting')\n",
    "        bam_conversion_command = f\"gsutil -u philipp-trollmann cat {bam_location} | samtools sort -n | bedtools bamtofastq -i - -fq {read1_fq} -fq2 {read2_fq}\"\n",
    "        subprocess.run(bam_conversion_command, shell=True, check=True)\n",
    "        print(f'Bam conversion for {cell_line} finished')\n",
    "\n",
    "        # Run Salmon quantification\n",
    "        print(f'Salmon quantification for {cell_line} starting')\n",
    "        salmon_command = f\"salmon quant -i {transcript_index} -l A -1 {read1_fq} -2 {read2_fq} --validateMappings -o {local_results_folder}\"\n",
    "        subprocess.run(salmon_command, shell=True, check=True)\n",
    "        print(f'Salmon quantification for {cell_line} finished')\n",
    "\n",
    "        # # Compress FASTQ files\n",
    "        # print(f'Compressing FASTQ files for {cell_line}')\n",
    "        # subprocess.run(f\"gzip {read1_fq} {read2_fq}\", shell=True, check=True)\n",
    "        # print(f'FASTQ files for {cell_line} compressed')\n",
    "        \n",
    "        # # Upload compressed FASTQ files to Google Cloud Storage\n",
    "        # output_folder = f\"{output_bucket}/cell_line_fq_data/{cell_line}/\"\n",
    "        # create_gcs_placeholder(output_folder)  # Create placeholder for FASTQ output folder\n",
    "        # subprocess.run(f\"gsutil -m cp -r {local_output_folder}*.gz {output_folder}\", shell=True, check=True)\n",
    "        # print(f'FASTQ files for {cell_line} uploaded to GCS')\n",
    "\n",
    "        # Upload Salmon quantification results to Google Cloud Storage\n",
    "        results_folder = f\"{output_bucket}/transcript_quant_results/{cell_line}/\"\n",
    "        create_gcs_placeholder(results_folder)  # Create placeholder for results folder\n",
    "        subprocess.run(f\"gsutil -m cp -r {local_results_folder}* {results_folder}\", shell=True, check=True)\n",
    "        print(f'Salmon quantification results for {cell_line} uploaded to GCS')\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error occurred during processing of {cell_line}: {e}\")\n",
    "        sys.exit(1)\n",
    "\n",
    "\n",
    "def main():\n",
    "    # Argument parser to accept cell line name and BAM location\n",
    "    parser = argparse.ArgumentParser(description=\"Process cell line and BAM location.\")\n",
    "    parser.add_argument(\"cell_line\", type=str, help=\"The cell line ID\")\n",
    "    parser.add_argument(\"bam_location\", type=str, help=\"The location of the BAM file in Google Cloud Storage\")\n",
    "    parser.add_argument(\"output_bucket\", type=str, help=\"The location of the google bucket to use\")\n",
    "    parser.add_argument(\"transcript_index\", type=str, help=\"The location of the local transcript index for the salmnon quantification\")\n",
    "    parser.add_argument(\"local_output_folder\", type=str, help=\"The location of the local tmp folder for the fq files\")\n",
    "    parser.add_argument(\"local_results_folder\", type=str, help=\"The location of the local tmp folder for the restuls files\")\n",
    "\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(args)\n",
    "\n",
    "    # Process the cell line and BAM location passed as arguments\n",
    "    process_cell_line(args.cell_line, args.bam_location, args.transcript_index, args.output_bucket, args.local_output_folder, args.local_results_folder)   \n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
